Enable matmul scheduler in segmenter

drzejan2 opened the pull request:
This is PR is a continuation of previous PR ([link](https://github.com/csarofeen/pytorch/pull/2461))

# The main goals of this PR are:

* update registry so segmenter can schedule matmul kernels,
* add an initial implementation of simple heuristics for kernel schedule details, but with reasonable partition to make it easier to extend functionality in future,
* add basic set of tests that will check new execution path,

# The side goals of this PR:
* organize matmul-related compile/runtime checks to avoid long build (avoiding extending `resgitry.cpp`)

# The compute-time check goals:

* start with simple checks to reject fusions that are obviously not supported
* when minimal requirements for matmul are met by fusion (number of fusion's input/outputs, type of these inputs) we proceed to more complex check that check dependencies in graph:
  * MmaOp inputs and relationship with fusion inputs:
    * MmaOp input must be a product of broadcast expression, that takes fusion input object as its object
  * MmaOp output and relationship with fusion output:
    * MmaOp output TensorView is fusion output

# Verification:
Platform: Ampere (GA100, 80GB), built with public CTK 11.8,
Branch sources: e4e0b2f77f964c9e767c64ab9f58d9a29239d534 (03.31.2023, after squashing)

### 1/ c++ tests
* command:
```
./build/bin/nvfuser_tests
```
* status, the failed tests on prepared branch:
```
// all tests passed
```

### 2/ torch tests
* command:
```
python third_party/nvfuser/python_tests/test_torchscript.py
```
* status, the failed tests on prepared branch (checked with current `main` and the results are the same):
```
----------------------------------------------------------------------
Ran 8221 tests in 1035.377s

FAILED (errors=153, skipped=8064)
```


### 3/ benchmarks
* command:
```
./build/bin/nvfuser_bench --benchmark_filter=NvFuserScheduler
```
* status, the failed tests on prepared branch:
```
// no issues found
```

drzejan2 commented:
There are the following issues to be addressed before this PR can be leave `draft` state:
1/ update handling of mma op in `scheduleMatmul` ([comment](https://github.com/csarofeen/pytorch/pull/2461#discussion_r1137093446))
2/ update pattern matching in compile-time checks ([comment](https://github.com/csarofeen/pytorch/pull/2461#discussion_r1137124134))

zasdfgbnm commented on csrc/scheduler/matmul_heuristic.h:
```diff
@@ -0,0 +1,147 @@
+    std::stringstream ss;
+    ss << "\n===== Matmul Parameters ========\n"
+       << (tag == "" ? "" : "Tag: ") << tag << "\n"
+       << "MMA op: " << nvfuser::toString(mma_op, true) << "\n"
```
nit: remove the `nvfuser::`

zasdfgbnm commented on csrc/scheduler/matmul_heuristic.h:
```diff
@@ -0,0 +1,147 @@
+  }
+
+  size_t hash() const override {
+    const size_t attr_hash = (nvfuser::hash(mma_op) << 0) //
```
Consider using https://github.com/pytorch/pytorch/blob/11114ab8bee2b0ddcb9f4d5a9378a5da32428ff3/c10/util/hash.h#L39

zasdfgbnm commented on csrc/scheduler/matmul_heuristic.h:
```diff
@@ -0,0 +1,147 @@
+      return false;
+    }
+
+    if (other_casted->layout != layout || //
```
nit: remove `//` at end of line

zasdfgbnm reviewed (commented):
Not done yet, posting some comments so far

drzejan2 commented on csrc/scheduler/matmul_heuristic.h:
```diff
@@ -0,0 +1,147 @@
+            rotate_ldmatrix_out_of_main_loop) {
+      return false;
+    }
+    if (!(other_casted->tile_sizes == tile_sizes) || //
+        !(other_casted->double_buffer_options == double_buffer_options)) {
```
I split the conditional expression into two as the separately checked members are complex structures that have own comparison operators. So first I check single-value members and then, if this check passes, I focused on complex structures.

But I agree with you, somehow the check became convolute. I will update it in the follow up commit.

Thanks!

drzejan2 commented:
@naoyam I checked what was the reason for failures caused by kernel index mode not being set.

Briefly: all requests for heuristics in existing schedulers set `params->cparams.index_type` in an explicit way. The major difference is that we use in these cases the following call:
```
indexModeToDtype(runtime_info.getIndexMode())
```

In the set of tools for matmul heuristics I added custom function to decide which index type use. While there is a known issue with `collectIndexMode` (does not take into consideration partial / output buffers shapes and right now I have no idea how to fix that), I update implementation of helper function to be a little bit more aware of possible pitfalls. My concern is that I rely there on tensor shapes (concrete domains extends) but I'm not sure if this is enough, I cannot find a good way to check strides per domain.

zasdfgbnm commented on csrc/mma_type.cpp:
```diff
@@ -238,4 +254,60 @@ std::string toString(MmaOptions::MacroType mt) {
+}
+
+size_t hash(const GemmTile& tile) {
+  return (std::hash<size_t>{}(static_cast<size_t>(tile.m)) << 0) //
```
> 'guide' automatic source formatting
That's a smart idea!

zasdfgbnm reviewed (commented):
Posting more comments, still not done

drzejan2 commented on test/test_gpu_tensorcore.cpp:
```diff
@@ -3060,6 +3039,96 @@ TEST_F(NVFuserTest, FusionAmpereMatmulLargeLoadLargeK_CUDA) {
 
+// Matmul test on Ampere relying on segmenter for 'C = A x B' fusion,
+//   with strict ref check hence single layout check
+TEST_F(NVFuserTest, FusionMatmulSegmenter_BasicMatmul_StrictCheck_TT_CUDA) {
```
I will update PR, thanks!

The background for initial naming convention - I wanted to emphasize key elements covered in prepared test, so:
- checks matmul scheduler through segmenter -> `FusionMatmulSegmenter`
- checks basic matmul pattern, `C = A x B` -> `BasicMatmul`
- the comparison is done with reference data with default checks, without loosing up expectations -> `StrictCheck`
- the test layout is `TT` only -> `TT`

drzejan2 commented on csrc/scheduler/matmul_heuristic.h:
```diff
@@ -0,0 +1,139 @@
+        other_casted->rotate_ldmatrix_out_of_main_loop ==
+        rotate_ldmatrix_out_of_main_loop &&
+        other_casted->tile_sizes == tile_sizes &&
+        other_casted->double_buffer_options == double_buffer_options;
```
Good catch - thanks! Will fix this with upcoming sources update.

drzejan2 commented on test/test_gpu_tensorcore.cpp:
```diff
@@ -3060,6 +3039,96 @@ TEST_F(NVFuserTest, FusionAmpereMatmulLargeLoadLargeK_CUDA) {
+
+    // NOTE: checking with lower expectations for relative/absolute error
+#if 1
+    TORCH_CHECK(outputs[0].allclose(tref, 0.001, 0.001));
```
Checked with `double` for `t0`, `t1` and `tref` - unfortunately, the detected error issue is till there.

drzejan2 commented on csrc/scheduler/matmul_utils.cpp:
```diff
@@ -0,0 +1,794 @@
+
+//! A helper for deciding what kernel indexing mode use (int32_t or int64_t).
+//! TODO: add strides to handle non-continous tensors
+PrimDataType getIndexType(const ProblemShape& problem_shape) {
```
It should be used, especially after change in implementation - will update the sources to use it again.

Thanks!

drzejan2 commented on csrc/scheduler/matmul_utils.cpp:
```diff
@@ -0,0 +1,794 @@
+  {
+    if (isAmpere(mma_op)) {
+      // Initial target:
+      // - 2 MMA ops per thread in a warp (32 threads), warp tile should be 64x
+      // instruction tile
+      constexpr DimType target_dim = 64;
+
+      const DimType m_ratio = target_dim / instruction_tile.m;
+      const DimType n_ratio = target_dim / instruction_tile.n;
+      const DimType k_ratio = target_dim / (m_ratio * n_ratio);
+
+      warp_tile = {
+          instruction_tile.m * m_ratio,
+          instruction_tile.n * n_ratio,
+          instruction_tile.k * k_ratio};
```
Perhaps I haven't named pieces of the algorithm in a clear way, sorry for that.

The general idea is to:

1.  have a warp tile, represented in elements, that will make each thread in warp (32 in total) have 2 mma ops  (it's for perf),
2. `target_dim` should be indeed named as `num_instructions` or `target_instructions_num`, as I targeted 64 instructions per warp,
3. I calculated ratios which will be used to calculate new `warp_tile` (in elements) based on mma op instruction shape,

for example, for mma instruction 161616 when we target 64 instructions per warp (2x instructions per thread) it will look like this:
```
// ratios:
m_ratio = 4 // 64/16
n_ratio = 4 // 64/16
k_ratio = 4 // 64/(4*4)

// warp tile shape
```
{
64, // result of instruction.m, 16, multiplied by 4
64, // result of instruction.n, 16, multiplied by 4
64, // result of instruction.k, 16, multiplied by 4
}
```
The warp tile shape can change due to:
- instruction shape (if we won't use wmma ops, e.g. 16816 on Ampere)
- adjust number of instructions per thread, instead of 2 we can have 1

Does it make sense?

zasdfgbnm commented on csrc/scheduler/matmul_utils.cpp:
```diff
@@ -0,0 +1,794 @@
+  const auto mma_inputs = mma_expr->inputs();
+  const auto mma_outputs = mma_expr->outputs();
+
+  // It is an unsupported fusion if
+  // - there are more than one fusion input TensorViews (producers)
+  //   for MMA op input
+  // - there are more than one fusion output TensorViews (consumers)
+  //   MMA op output
+  const auto getKeyTvFromPathBetween =
+      [](const std::vector<Val*>& producers,
+         const std::vector<Val*>& consumers) -> Val* {
+    const auto paths = getAllDepndencyChains(producers, consumers);
+
+    if (paths.empty()) {
+      return nullptr;
+    }
+
+    std::vector<Val*> tvs;
+    for (const auto& path : paths) {
+      if (path.empty()) {
+        continue;
+      }
+      if (path.front()->isA<TensorView>()) {
+        tvs.push_back(path.front());
+      }
+    }
+    return (tvs.size() == 1) ? tvs[0] : nullptr;
+  };
+
+  const auto* tv_input_A =
+      getKeyTvFromPathBetween(fusion_inputs, {mma_inputs[0]});
+  if (nullptr == tv_input_A) {
+    return c10::nullopt;
+  }
+
+  const auto* tv_input_B =
+      getKeyTvFromPathBetween(fusion_inputs, {mma_inputs[1]});
+  if (nullptr == tv_input_B) {
+    return c10::nullopt;
+  }
```
I think it makes sense. I didn't finish reading yet, and didn't know we already start thinking about simple fusions.

zasdfgbnm commented on csrc/scheduler/matmul_utils.cpp:
```diff
@@ -0,0 +1,794 @@
+  }
+
+  // set kernel index mode
+  params->cparams.index_type = getIndexType(problem_shape.value());
```
Hmmm, all the other schedulers are using  `indexModeToDtype(runtime_info.getIndexMode());`, if matmul is having trouble with it, then very likely all other schedulers are also having trouble with it. I am OK with your `getIndexType` approach today, but I think I should investigate this issue deeper.

zasdfgbnm replied:
Yes, I was right. Other schedulers are also affected: https://github.com/NVIDIA/Fuser/issues/79

drzejan2 commented:
The scope of latest change:
- rebased to `main`
- updated code to handle cta tile rasterization order flag, both in matmul heuristic params as well as in `scheduleMatmul` (based on #76)
- fixed several of issues detected with `clang-tidy`

I redone verification with latest sources, no red flags found.

@zasdfgbnm, @naoyam and @csarofeen for visibility

zasdfgbnm commented on csrc/scheduler/matmul_utils.cpp:
```diff
@@ -0,0 +1,795 @@
+
+  size_t status = 0;
+
+  const auto filterTvsFromSet = [](const std::vector<Val*>& source) {
```
I think this is the same as `ir_utils::filterByType<TensorView>(source)`

zasdfgbnm commented on csrc/scheduler/matmul.cpp:
```diff
@@ -46,14 +53,35 @@ void moveInnerBroadcastLeft(TensorView* tv, int number_of_inner_pos = 3) {
+      outputs[0]->isA<TensorView>(),
+      "fusion's output is not an instance of TensorView class");
+
+  TensorView* a = inputs[0]->as<TensorView>();
+  TensorView* b = inputs[1]->as<TensorView>();
+  TensorView* c = outputs[0]->as<TensorView>();
```
This PR has some inconsistency here, some part of this PR tries to allow arbitrary order of fusion inputs and fusion outputs, but in `scheduleMatmul`, we are still hard coding the order. Should we have a unified scheme on how to get ABC? For example, we can have

```C++
auto mma = findMMAOp(fusion);
auto a = getAFromMMA(mma);
auto b = getBFromMMA(mma);
auto c = getCFromMMA(mma);
```

zasdfgbnm replied:
Also, I think we need to add a test to check if different order are actually supported

drzejan2 replied:
In this PR I focused on adding matmul scheduler and plugging in `schedulMatmul` to the system / segmenter. All the checks in segmenter are made so it should be reasonable to add new rules for checks when new functionality is added. This is the reason why latest revision supports an arbitrary order of inputs/outputs in the fusion - this is preparation for the support for more complex fusions.

On the other hand, I tried to avoid doing any changes in `scheduleMatmul` as the scope of this PR is already big (plugging in scheduler, compile time checks and heuristics). 

---

As for your suggestions about helpers / tools that for given fusion, will return requested input of mma from fusion. I have in my mind the following idea:
- compile time checks performs heavy analysis of dependencies in fusion:
  - if there is enough TV inputs in fusion to have at least the basic mma (2 TVs), 
  - if no more than single fusion TV input is a producer for MmaOp input
  - if there are scalar inputs (alpha/beta)
  - if there are additional TVs (out-of-place C for beta scaling)
- the product of compile-time checks are the following pieces information, that could be stored in data cache (instance of `HeuristicSummary`):
  - mapping fusion inputs / inputs' consumers to their roles in expressions in fusion, i.e. for `(alpha * A) x B` we will mark consumer of `alpha * A` as the first input for MmaOp.
  - store mma layout
- instance of heuristics params passed to `scheduleMarmul` will contain this data so we can trigger process piece by piece there. 

The major issues is that instance of `HeuristicSummary` can be initialized starting from runtime check. The stopgap solution for now would be to redo heavy analysis from compile-time check in runtime check, before any major changes are done to projects' architecture.

---

> Also, I think we need to add a test to check if different order are actually supported

I prepared a simple test like this, but without access to data from compile-time about mapping fusion inputs to mma inputs it failed in `scheduleMatmul`, on checks you have pointed out. Do you mind if implementation and test is added in the follow up PR?

zasdfgbnm commented on csrc/scheduler/matmul_utils.cpp:
```diff
@@ -0,0 +1,795 @@
+    // this means there are no additional operations in between
+    {
+      for (const auto& path : paths_mma_C_to_outputs) {
+        if (path.size() != 1) {
```
I am seeing you grabbing size 1 paths a lot. What's the benefit of grabbing such a path, compared with just the following?
```C++
if (!mma->output(0)->isFusionOutput()) {
  return {c10::nullopt, "One of paths between MMA op output and fusion outputs contains a product of unsupported operation"};
}
```
Are you thinking about potential fusion?

drzejan2 commented on csrc/scheduler/matmul_heuristic.h:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    size_t hash() const {
+      return (std::hash<size_t>{}(static_cast<size_t>(double_buffer_smem_write))
+              << 0) ^
+          (std::hash<size_t>{}(static_cast<size_t>(double_buffer_smem_read))
+           << 1) ^
+          (std::hash<size_t>{}(static_cast<size_t>(smem_double_buffer_stage))
+           << 2);
```
For this I followed the example provided to `std::hash` documentation on [this page](https://en.cppreference.com/w/cpp/utility/hash).

Do you think the current implementation might be slower / has a higher chance of hash conflicts?

csarofeen replied:
Also curious about this, why `&` @zasdfgbnm ?

drzejan2 commented on csrc/scheduler/registry.cpp:
```diff
@@ -2054,10 +2055,67 @@ class PersistentKernelScheduler : public SchedulerEntry {
+  }
+
+  static bool canScheduleCompileTime(Fusion* fusion) {
+    // Check that inputs of all select/gather-like ops are fusion inputs
+    if (rejectScheduleForSelectLikeOps(fusion, ScheduleHeuristic::Matmul)) {
+      return false;
+    }
+
```
That's a good point - I though this is a common requirement for any of schedulers defined in registry.

So it is ok to remove this check and assume, for now, that none of expressions, such as `SelectOp` / `IndexSelectOp` / `TorchGatherOp`, is present in prepared fusion? Then I can add new rule to compile-time checks, to explicitly look for these expr and reject scheduling fi any found?

zasdfgbnm replied:
Because we are not doing any fusion right now, so there will automatically be no `select/index_select/gather`.

drzejan2 commented:
The scope of latest changes in the PR:

- only `matmul scheduler` can process fusions that have MmaOp expr, others will reject processing fusions,
- simplify hashing functions used in matmul heuristic params (flags are combined and then hashed)
- MmaOp expression constructor will now perform check if provided inputs contain broadcast domains (previously defined as a rule in compile time checks in matmul scheduler)
- remove check for memory location of fusion inputs/outputs and MmaOp inputs from rules in compile time checks for matmul scheduler (fusion inputs/outputs are, by definition, always in global memory; MmaOp inputs can be redefined based on scheduler decisions),
- simplified dependency check for MmaOp inputs and outputs:
  - inputs: must be a product of Broadcast expression, which takes as input one of fusion inputs
  - output: must be a fusion output
- simplified rules for fusion inputs/outputs in compile time checks for matmul scheduler
  - explicit expectation for the number of inputs (2) and outputs (1)
  - inputs/outputs can be only TensorViews

The current status of verification: restarted all tests, I will update PR description when those are done.

@zasdfgbnm and @naoyam for visibility of applied changes.

mmigdal-nv commented on csrc/scheduler/matmul_heuristic.h:
```diff
@@ -0,0 +1,160 @@
+      return ss.str();
+    }
+
+    size_t hash() const {
```
Very naive, but why not going with `std::hash<>` specializations instead ?

zasdfgbnm replied:
I agree. We should change that (but not in this PR), not only for matmul, but for all schedulers.

drzejan2 replied:
Such specialization will require extending namespace `std` that I'm a little bit hesitant to do (although this is the way we are supposed to do it, if I'm not wrong). Maybe it's not a bad idea and we could have all matmul related structs, that require hashing, have own specialization - this can be checked and done as a follow up of this PR. Does it sound ok?

zasdfgbnm reviewed (commented):
Left a few more comments.

zasdfgbnm commented on csrc/ir_nodes.cpp:
```diff
@@ -1351,6 +1351,17 @@ MmaOp::MmaOp(
           in_b->getValType().value() == ValType::TensorIndex,
       in_b->getValType().value());
 
+  const auto isBroadcastIn = [](const Val* val) {
+    if (val->getValType().value() == ValType::TensorView) {
+      const auto* tv = val->as<TensorView>();
+      return tv->hasBroadcast();
+    }
+    return true;
+  };
+
+  TORCH_INTERNAL_ASSERT(isBroadcastIn(in_a));
+  TORCH_INTERNAL_ASSERT(isBroadcastIn(in_b));
```
Note for the future:

I think we should not only check the input has a broadcast, but also to add many more consistency check. For this PR, I think it is fine, but let's create a separate followup PR for it:

I think we should check:
- Inputs has two concrete IDs and one broadcast ID
- The broadcast's axis in different inputs are different
- Output has two concrete IDs and one reduction ID
- The axis of the output reduction ID must correspond to a concrete ID in both inputs

And with this check here, we can remove corresponding checks in the scheduler.

zasdfgbnm commented on csrc/scheduler/matmul_utils.cpp:
```diff
@@ -0,0 +1,634 @@
+
+//! A helper for checking if layout of MMA op's inputs. It will return optional
+//! message if check fails.
+LayoutData getInputsLayout(const MmaOp* mma_expr) {
```
Note for future PR:
Does it make sense to make this a method of `MmaOp`?

drzejan2 replied:
You are right, this can be moved as method of MmaOp. I will prepare PR with this change.

Thanks!

zasdfgbnm reviewed (approved):
I don't think there is any blockers, and this PR is good enough to get merged. Considering that this PR conflicts with lots of other matmul schedule changes such as NN support, split-k support, etc. I think the best strategy is to merge this PR early, and we can always do further improvements in followup PRs.

drzejan2 commented on csrc/scheduler/matmul_utils.cpp:
```diff
@@ -0,0 +1,634 @@
+    }
+
+    // Fusion can only have TVs as outputs, and there can be only one output
+    if (fusion_outputs_tvs.size() != fusion_outputs.size()) {
+      return "Fusion has output which is not a TensorView object";
+    }
```
Good to know - in the follow up PR, with code cleaning, I will remove this check in `matmul_utils.cpp`.

Thanks for letting me know about executor's requirement.

drzejan2 commented on csrc/scheduler/matmul.cpp:
```diff
@@ -7,8 +7,15 @@
+// 'SchedulerRuntimeInfo'
+#include <executor_utils.h>
+
+#include <sstream>
```
It looks like an artifact from one of the older revisions of compile /run time checks (to generate reason message for check failure). Will remove it in the follow up PR.

Thanks!
